### PR TITLE
Add annotations

### PR DESCRIPTION
**Pull request by apdibbo**
Various changes to make the template library more useable and more validatable:
Create intermediary dicts for database, rabbitmq and memcached config
Use functions to build database connection strings and rabbitmq and memcached server lists
Move intermediary variables to another template so that config just has the variables that should be set by the user
Add Annotations to all global variables in defaults/openstack/config
Add schemas and types for Openstack component ini files (just the *.conf files so far) just the commonality
Bind schemas to metaconfig paths

### CODE REVIEW

File: defaults/openstack/functions.pan [hunk 1]
```diff
@@ -43,4 +43,52 @@ function openstack_list_to_string = {
...
+
+  if (is_dict(ARGV[0])) {
+    config = ARGV[0];
+  } else {
+    error('openstack_dict_to_hostport_string needs a dict as an argument');
+  };
+  result = '';
+  foreach(k;v;config) {
+    if (result != '') {
+      result = result + ',' + k + ':' + v;
```
Comment (guillaume-philippon) [inline defaults/openstack/functions.pan]: Can't we use format('%s,%s:%d', result, k, v) ?

File: defaults/openstack/config.pan [hunk 2]
```diff
@@ -4,48 +4,126 @@ unique template defaults/openstack/config;
 # Define site specific variables #
 ##################################
 include if_exists('site/openstack/config');
+@use{
+  type = ipaddress
+  note = This is the primary IP address of the Host
+}
 variable PRIMARY_IP ?= DB_IP[escape(FULL_HOSTNAME)];
```
Comment (guillaume-philippon) [inline defaults/openstack/config.pan:11]: Is there any reason to not rename it OPENSTACK_PRIMARY_IP ? to match variable name convention ?

File: defaults/openstack/functions.pan (hunk 1 above)
Comment (apdibbo) [reply defaults/openstack/functions.pan]: No reason not to, I have just pushed a commit to correct this

File: defaults/openstack/config.pan (hunk 2 above)
Comment (apdibbo) [reply defaults/openstack/config.pan:11]: PRIMARY_IP is a construct that exists in Aquilon, this line is there to provide backwards compatibility to SCDB

**Comment by apdibbo** [comment]
I have also pulled the related changes from liberty in here

File: features/accounts/config.pan [hunk 3]
```diff
@@ -47,3 +47,6 @@ prefix '/software/components/accounts';
 # MongoDB user
 'kept_users/mongod' = '';
 'kept_groups/mongod' = '';
```
Comment (guillaume-philippon) [inline features/accounts/config.pan:49]: Should not be on this PR as is discuss on PR#32...

File: defaults/openstack/config.pan (hunk 2 above)
Comment (guillaume-philippon) [reply defaults/openstack/config.pan:11]: Oh, yes, I renember now.

**Comment by guillaume-philippon** [comment]
It seems good for me. Do you have deployed it @RAL ? If so we can merge it.

**Comment by apdibbo** [comment]
Good, I do indeed have this deployed at RAL

**Pull request merged**